src/resources/uart_esp32.cc

```diff
@@ -119,135 +608,230 @@ PRIMITIVE(init) {
   return proxy;
 }
 
+typedef struct {
+  QueueHandle_t queue;
+  uart_hal_handle_t hal;
+  bool hardware_initialized;
+  UartResource* uart_resource;
+  uart_port_t port;
+  uint8_t* rx_buffer;
+  uint8_t* tx_buffer;
+} uart_initialization_record_t;
```

Comment:
I think it's better to make this a class (with public fields) and convert uart_init_cleanup to a destructor that does not have to be called explicitly.